Add bot field to certificates and various usage events

timothyb89 opened the pull request:
This adds a new certificate extension field, `teleport-bot`, to certificates issued to Machine ID bot users that can definitively identify certificates as having been issued to a bot user, set to the bot's unprefixed name.

Additionally, this uses the new bot name field to mark certain usage events as originating from bot users. As such, it includes a protobuf update from Cloud [1], which also pulled in some small additional (mostly comment) changes.

[1] https://github.com/gravitational/cloud/pull/7060

changelog: Added new certificate extensions and usage reporting flags to explicitly identify Machine ID bots and their cluster activity

timothyb89 committed:
Add bot field to certificates and various usage events

This adds a new certificate extension field, `teleport-bot`, to
certificates issued to Machine ID bot users that can definitively
identify certificates as having been issued to a bot user.

Additionally, this uses the new `Bot` identity flag to mark certain
usage events as originating from bot users. As such, it includes a
protobuf update from Cloud [1], which pulled in some small additional
(mostly comment) changes.

[1] https://github.com/gravitational/cloud/pull/7060

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-flag-usage-events

timothyb89 commented:
Outstanding TODOs:

- [ ] Manually test the event pipeline (needs to emit events with the flag set as expected)
- [ ] Unit tests for the bot cert flag

timothyb89 committed:
Small bot flag plumbing fixes

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-flag-usage-events

timothyb89 commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -92,6 +92,9 @@ message UserMetadata {
   string RequiredPrivateKeyPolicy = 9 [(gogoproto.jsontag) = "required_private_key_policy,omitempty"];
+
+  // Bot indicates this user is a Machine ID bot user.
+  bool Bot = 10 [(gogoproto.jsontag) = "bot,omitempty"];
```
After only wasting a day debugging why bot events are always being reported as `human` rather than `unspecified` for outdated nodes... this needs to be changed to an enum as used in the prehog protos to account for outdated clients.

timothyb89 replied:
Alright, this appears to be fixed now, verified on a Cloud tenant with both a vanilla v13 node which produces `user_kind: ""` aka unspecified and a node running this branch which correctly indicates bot/human.

strideynet commented on lib/auth/keygen/keygen.go:
```diff
@@ -206,6 +206,9 @@ func (k *Keygen) GenerateUserCertWithoutValidation(c services.UserCertParams) ([
 		cert.Permissions.Extensions[teleport.CertExtensionGeneration] = fmt.Sprint(c.Generation)
 	}
+	if c.Bot {
+		cert.Permissions.Extensions[teleport.CertExtensionBot] = ""
```
Maybe we should put the name of the Bot into the extension rather than just setting the extension - I figure it might be useful eventually?

timothyb89 replied:
Good thought, I've converted the embedded cert field to a string containing the bot name (and renamed the field where relevant).

timothyb89 committed:
Convert bot flag to BotName and UserKind enum

This makes a few changes to the bot tagging approach:
* The bot name is embedded in the cert rather than just true/false
* UserKind is included in events rather than just a bot flag, to
  allow for an unspecified value for older client nodes.

timothyb89 committed:
Add a quick unit test for bot cert extensions

timothyb89 committed:
Fix outdated grpc

strideynet commented on lib/usagereporter/teleport/aggregating/reporter.go:
```diff
@@ -211,10 +224,14 @@ func (r *Reporter) run(ctx context.Context) {
 	userActivity := make(map[string]*prehogv1.UserActivityRecord)
 
-	userRecord := func(userName string) *prehogv1.UserActivityRecord {
+	userRecord := func(userName string, v1AlphaUserKind prehogv1alpha.UserKind) *prehogv1.UserActivityRecord {
```
Nit: I don't think it's a major concern but I can see a bit of potential raciness here if there's an issue with tracking bot status. The status of the user from the first event will set it for all future events. Perhaps we ought to put some warning or error in if the kind of the user changes ?

timothyb89 replied:
I added a warning and had it override the flag. I think the behavior could go either way realistically - and given it's a bug regardless I'm not sure if I expect anything good to come from whichever behavior we use.

espadolini replied:
Maybe we should leave the `record.UserKind` as is if `v1UserKind` is unspecified, just in case?

timothyb89 replied:
I've tweaked the logic here. It now allows identifying a previously unspecified user, ignores unspecified updates, and warns about flipping between human/bot users.

strideynet reviewed (commented):
Looks generally good. From my reading, we're only currently putting the attribute into the impersonated certs or renewed certs. Could we also make sure it goes onto the bots own certs that are produced by the joining process ?

timothyb89 committed:
Include bot flag on initial certs

timothyb89 committed:
Log a warning and override user kind for usage records if they differ

timothyb89 commented:
> From my reading, we're only currently putting the attribute into the impersonated certs or renewed certs. Could we also make sure it goes onto the bots own certs that are produced by the joining process ?

Good catch on the bot flag, I always forget the initial certs are generated from a different codepath. The flag is fixed now.

timothyb89 committed:
Fix several unit tests; add a bot metadata test case

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-flag-usage-events

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-flag-usage-events

timothyb89 committed:
Fix unit tests with UserKind zero value

espadolini commented on proto/prehog/v1alpha/teleport.proto:
```diff
@@ -657,6 +685,9 @@ message KubeRequestEvent {
   string user_name = 1;
+
+  // Indicates the kind of user (e.g. bot or human), if known.
+  UserKind user_kind = 2;
```
Are we sure we don't want a string here? We seem to be using strings (for the most part) for enum-y things here in the streamed events.

espadolini commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -63,6 +63,21 @@ message SessionMetadata {
+
+  // Indicates the user associated with this event is human, either created
+  // locally or via SSO.
+  USER_KIND_HUMAN = 1;
```
Is "human" really the correct angle here? Could we make it a potentially more useful "login"/"tctl"/"bot"?

timothyb89 committed:
Rename SSH cert extension to use standard format

Renames the `teleport-bot` extension to `bot-name@goteleport.com`,
to better follow SSH cert extension naming conventions.

timothyb89 committed:
Attempt to improve unspecified userkind aggregating logic

timothyb89 commented on constants.go:
```diff
@@ -495,6 +495,9 @@ const (
 	CertExtensionDeviceCredentialID = "teleport-device-credential-id"
+	// CertExtensionBot indicates the name of the Machine ID bot this
+	// certificate was issued to, if any.
+	CertExtensionBot = "teleport-bot"
```
Fair enough, fixed!

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/bot-flag-usage-events

timothyb89 committed:
Merge branch 'master' into timothyb89/bot-flag-usage-events

timothyb89 committed:
Merge branch 'master' into timothyb89/bot-flag-usage-events
